File: project/backend-services/smart-irrigation-backend/domain-services/src/main/java/pt/sensae/services/smart/irrigation/backend/domainservices/irrigationZone/mapper/IrrigationZoneMapper.java
package pt.sensae.services.smart.irrigation.backend.domainservices.irrigationZone.mapper;

import pt.sensae.services.smart.irrigation.backend.domain.model.GPSPoint;
import pt.sensae.services.smart.irrigation.backend.domain.model.business.irrigationZone.*;
import pt.sensae.services.smart.irrigation.backend.domainservices.irrigationZone.model.BoundaryCommandDetails;
import pt.sensae.services.smart.irrigation.backend.domainservices.irrigationZone.model.CreateIrrigationZoneCommand;
import pt.sensae.services.smart.irrigation.backend.domainservices.irrigationZone.model.UpdateIrrigationZoneCommand;

import java.util.List;
import java.util.stream.Collectors;

public class IrrigationZoneMapper {

    public static IrrigationZone toModel(UpdateIrrigationZoneCommand updateCommand) {
        var of = IrrigationZoneId.of(updateCommand.id);
        return getGardeningArea(updateCommand.area, of, updateCommand.name);
    }

    public static IrrigationZone toModel(CreateIrrigationZoneCommand createCommand) {
        var of = IrrigationZoneId.create();
        return getGardeningArea(createCommand.area, of, createCommand.name);
    }

    private static IrrigationZone getGardeningArea(List<BoundaryCommandDetails> boundaryCommands, IrrigationZoneId of, String name) {
        var area = new Area(boundaryCommands.stream().map(b -> {
            var gps = new GPSPoint(b.latitude, b.longitude, b.altitude);
            return new BoundaryPoint(b.position, gps);
        }).collect(Collectors.toSet()));

        return new IrrigationZone(of, IrrigationZoneName.of(name), area, Owners.empty());
    }
}

File: project/frontend-services/libs/smart-irrigation/mapper/src/lib/mappers/IrrigationZoneMapper.ts
import {IrrigationZoneDTO} from "@frontend-services/smart-irrigation/dto";
import {
  AreaBoundary,
  IrrigationZone,
  IrrigationZoneId,
  IrrigationZoneName,
  GPSDataDetails
} from "@frontend-services/smart-irrigation/model";

export class IrrigationZoneMapper {
  static dtoToModel(dto: IrrigationZoneDTO): IrrigationZone {
    const area = dto.area.map(b => new AreaBoundary(b.position, new GPSDataDetails(b.latitude, b.longitude, b.altitude)));
    return new IrrigationZone(new IrrigationZoneId(dto.id), new IrrigationZoneName(dto.name), area);
  }
}
